Add Turris Omnia MCU driver + firmware + tool

elkablo opened the pull request:
Add Turris Omnia MCU kernel driver, MCU firmware package and the `omnia-mcutool` utility:
- the driver is backported to 6.6. It is backported from Linus' master branch (will be released in 6.11)
- the firmware comes from [here](https://gitlab.nic.cz/turris/hw/omnia_hw_ctrl/-/releases/v4.0)
- the `omnia-mcutool` utility comes from [here](https://gitlab.nic.cz/turris/omnia-mcutool)

The MCU driver adds the following abilities if the user has upgraded MCU firmware (see below as to how to do that):
- the poweroff command now completely powers the device off (the power consumption goes to minimum). To power the router on, the user needs to press the front button,
- the device can be configured to automatically power on from powered off state at a specific time by using the `omnia-mcutool --wakeup` option, for example
    ```
    omnia-mcutool --wakeup '+10 hours'
    ```
  or
    ```
    omnia-mcutool --wakeup '2023-10-19 08:00'
    ```

The following features are supported by the MCU driver even without upgrading the MCU firmware:
- the factory reset level can be read by
    ```
    cat /sys/bus/i2c/devices/1-002a/reset_selector
    ```
- the front button can be configured to act as an input device (via the
  `kmod-input-gpio-keys` driver). This is done by executing
    ```
    echo cpu >/sys/bus/i2c/devices/1-002a/front_button_mode
    ```
  To configure the default option (pressing front button dimms the front
  LEDs panel), execute
    ```
    echo mcu >/sys/bus/i2c/devices/1-002a/front_button_mode
    ```
- some of the MCU GPIOs are now available via `/sys/class/gpio`.
  For example we can now enable/disable power for USB ports via sysfs:
    ```
    cd /sys/class/gpio
    echo 456 >export
    cd '/sys/class/gpio/gpio456.Front USB3 port power'
    ```
  Now to disable USB port power
    ```
    echo 0 >value
    ```
  and to enable it back
    ```
    echo 1 >value
    ```

The user can upgrade the MCU firmware by executing
  ```
  opkg update
  opkg install omnia-mcutool
  omnia-mcutool --upgrade
  reboot
  ```

lixkel committed:
mvebu: Add kmod-turris-omnia-mcu

Add support for the MCU driver on CZ.NIC's Turris Omnia. This adds
the ability to do a true board poweroff, and to configure various
features (for example the user may configure that after poweroff, the
router should automatically wake up at a specific time).

Signed-off-by: Marek Mojík <marek.mojik@nic.cz>
Signed-off-by: Marek Behún <kabel@kernel.org>
Link: https://github.com/openwrt/openwrt/pull/13799
Signed-off-by: Robert Marko <robimarko@gmail.com>

lixkel committed:
firmware: Add CZ.NIC Turris Omnia MCU firmware

Add a new package, omnia-mcu-firmware, containing firmware binaries for
the microcontroller on the Turris Omnia router.

Signed-off-by: Marek Mojík <marek.mojik@nic.cz>
Signed-off-by: Marek Behún <kabel@kernel.org>
Link: https://github.com/openwrt/openwrt/pull/13799
Signed-off-by: Robert Marko <robimarko@gmail.com>

lixkel committed:
utils: Add the omnia-mcutool utility

Add a new utility, omnia-mcutool, which main purpose is to upgrade the
firmware on the microcontroller on the Turris Omnia router. Depends on
omnia-mcu-firmware, and the upgrade process is pretty simple:

  omnia-mcutool --upgrade

Besides firmware upgrade, the utility can be used to show and configure
various firmware settings.

Signed-off-by: Marek Mojík <marek.mojik@nic.cz>
Signed-off-by: Marek Behún <kabel@kernel.org>
Link: https://github.com/openwrt/openwrt/pull/13799
Signed-off-by: Robert Marko <robimarko@gmail.com>

hauke commented on target/linux/mvebu/modules.mk:
```diff
@@ -17,6 +17,27 @@ endef
+define KernelPackage/turris-omnia-mcu
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=CZ.NIC's Turris Omnia MCU driver
+  DEPENDS:=@TARGET_mvebu_cortexa9 @TESTING_KERNEL
```
Please use `@!LINUX_5_15`instead of `@TESTING_KERNEL`. 

We could miss this when we change mvebu to kernel 6.1 by default. when you use not 5.15 we will find it when we remove kernel 5.15 support and can remove the dependency.

elkablo replied:
I backported for 5.15 as well, so no need to this. Will push once I test it.

hauke commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+define Package/omnia-mcu-firmware
+  SECTION:=firmware
+  CATEGORY:=Firmware
+  URL:=$(PKG_SOURCE_URL)
```
This should point to the project URL, this looks good: https://gitlab.nic.cz/turris/hw/omnia_hw_ctrl/-/releases

elkablo replied:
ok

hauke commented on target/linux/mvebu/modules.mk:
```diff
@@ -17,6 +17,27 @@ endef
+   CONFIG_CZNIC_PLATFORMS=y \
+   CONFIG_TURRIS_OMNIA_MCU
+  FILES:=$(LINUX_DIR)/drivers/platform/cznic/turris-omnia-mcu.ko
+  AUTOLOAD:=$(call AutoLoad,59,turris-omnia-mcu,1)
```
We normally prefer ÀutoProbe` over `AutoLoad`, Please try if it works too.

elkablo replied:
ok

hauke commented:
I didn't look into the kernel driver. If it gets accepted please update the drover to the version which was accepted.

BKPepe commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_NAME:=omnia-mcu-firmware
+PKG_DISTNAME:=omnia_hw_ctrl
+PKG_VERSION:=3.4
+PKG_RELEASE:=$(AUTORELEASE)
```
Unfortunately, autorelease was deprecated here, though. :-( It was handy feature and I am sad every time when someone forgets to bump the version.

```suggestion
PKG_RELEASE:=1
```

elkablo replied:
ok

BKPepe commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+endef
+
+define Package/omnia-mcu-firmware/description
+Firmware binaries for the microcontroller on the Turris Omnia router. These are
+used by the omnia-mcutool utility when upgrading MCU firmware.
```
```suggestion
  Firmware binaries for the microcontroller on the Turris Omnia router. These are
  used by the omnia-mcutool utility when upgrading MCU firmware.
```

elkablo replied:
See my reply to this suggestion for `omnia-mcutool`.

BKPepe commented on package/utils/omnia-mcutool/Makefile:
```diff
@@ -0,0 +1,49 @@
+
+PKG_NAME:=omnia-mcutool
+PKG_VERSION:=0.2
+PKG_RELEASE:=$(AUTORELEASE)
```
```suggestion
PKG_RELEASE:=1
```

elkablo replied:
ok

BKPepe commented on package/utils/omnia-mcutool/Makefile:
```diff
@@ -0,0 +1,49 @@
+endef
+
+define Package/omnia-mcutool/description
+The omnia-mcutool utility is mainly used to upgrade the firmware on the
+microcontroller on the Turris Omnia router. It can also show state of MCU
+settings and configure MCU options (GPIOs, LEDs, power).
```
```suggestion
  The omnia-mcutool utility is mainly used to upgrade the firmware on the
  microcontroller on the Turris Omnia router. It can also show state of MCU
  settings and configure MCU options (GPIOs, LEDs, power).
```

elkablo replied:
I disagree, because then in `make menuconfig` it looks like this
![menuconfig](https://github.com/openwrt/openwrt/assets/1883885/27c1e59f-1b57-4aae-884d-d0a3f2386e93)

Notice that the indentation does not take on the first line, only from the second one. It seems that it does not work correctly for multi-line descriptions.

Looking at the tree with `git grep -A 2 'Package/.*/desc', there are some packages with multi-line descriptions that do indent (like `kobs-ng`, `gperf`) and some that do not (`gdb`, `gdbserver`).

BKPepe commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_HASH:=2b0888fb065b279dc5eac90a76fea8e2353bd554b16df07515c4114d6d0a2c8e
+
+PKG_MAINTAINER:=Marek Mojik <marek.mojik@nic.cz>
+PKG_LICENSE:=GPL-3.0-or-later
```
```suggestion
PKG_LICENSE:=GPL-3.0-or-later
PKG_LICENSE_FILES:=LICENSE
```

elkablo replied:
What if the `LICENSE` file is not contained in the `$(PKG_SOURCE_URL)` tarball? That tarball only contains built binaries. Will this be correct?

BKPepe commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+
+define Package/omnia-mcu-firmware/install
+	$(INSTALL_DIR) $(1)/usr/share/omnia-mcu-firmware/
+	$(INSTALL_DATA) $(PKG_BUILD_DIR)/*.bin $(1)/usr/share/omnia-mcu-firmware/
```
Can not we specify which binaries we will have in the package instead of having all binaries, there?

elkablo replied:
I don't know, I have specifically created the tarball (`omnia_hw_ctrl-binaries-v3.4.tar.bz2`) to contain only the files that need to be installed, so that I would not need to manually name the 10 files in the package Makefile :)

Grepping through the tree, there are several packages using file name expansions in install macros, `git grep 'INSTALL.*/\*'` matches 64 lines in the repository.

BKPepe commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_VERSION:=3.4
+PKG_RELEASE:=$(AUTORELEASE)
+
+PKG_SOURCE_SUBDIR:=$(PKG_DISTNAME)-binaries-v$(PKG_VERSION)
+PKG_SOURCE:=$(PKG_SOURCE_SUBDIR).tar.bz2
```
I am thinking. Do we really need to have extra variable - ``PKG_DISTNAME``? Can not we have it only as ``omnia-hw-ctrl``?

elkablo replied:
I was inspired by package `u-boot-envtools`... Notice that the variable is then used at 3 places.

BKPepe commented on package/firmware/omnia-mcu-firmware/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_SOURCE_URL:=https://gitlab.nic.cz/turris/hw/$(PKG_DISTNAME)/-/releases/v$(PKG_VERSION)/downloads/
+PKG_HASH:=2b0888fb065b279dc5eac90a76fea8e2353bd554b16df07515c4114d6d0a2c8e
+
+PKG_MAINTAINER:=Marek Mojik <marek.mojik@nic.cz>
```
Is other Marek aware that he will take care about this package in upstream?

elkablo replied:
Of course :) And if he changes his mind, I will take over.

BKPepe commented on package/utils/omnia-mcutool/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_RELEASE:=$(AUTORELEASE)
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
+PKG_SOURCE_URL:=https://gitlab.nic.cz/turris/omnia-mcutool/-/archive/$(PKG_VERSION)/
```
Could you add LICENSE to your repository, please?

BKPepe commented on package/utils/omnia-mcutool/Makefile:
```diff
@@ -0,0 +1,49 @@
+define Package/omnia-mcutool
+  SECTION:=utils
+  CATEGORY:=Utilities
+  URL:=$(PKG_SOURCE_URL)
```
Same as @hauke pointed out here.

elkablo replied:
ok

hauke commented on target/linux/generic/pending-6.1/816-1-dt-bindings-arm-add-cznic-turris-omnia-mcu-binding.patch:
```diff
@@ -0,0 +1,99 @@
+ create mode 100644 Documentation/devicetree/bindings/arm/cznic,turris-omnia-mcu.yaml
+
+--- /dev/null
++++ b/Documentation/devicetree/bindings/arm/cznic,turris-omnia-mcu.yaml
```
Do you plan to use this driver on other targets too? If not the patches should go into target/linux/mvebu/patches-6.1. 
If you keep it here please add the refreshing of the other patches also into this patch which broke them.

elkablo replied:
OK, I'll move it to mvebu.

elkablo commented:
@hauke, @BKPepe did not comment on my answers to the 2 unresolved review entries. Should I mark them resolved?

BKPepe commented:
Ehm... wasn't there weekend? 😅 No need to hurry I guess.

Dne po 30. 10. 2023 16:49 uživatel Marek Behun ***@***.***>
napsal:

> @hauke <https://github.com/hauke>, @BKPepe <https://github.com/BKPepe>
> did not comment on my answers to the 2 unresolved review entries. Should I
> mark them resolved?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/13799#issuecomment-1785509712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7IDVEMVW4CGEVZTG2ZOTTYB7EB3AVCNFSM6AAAAAA6RSRW56VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOBVGUYDSNZRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

elkablo commented:
@BKPepe if you could find time to answer the two questions I would be glad.

stklcode commented:
Unfortunately I do not have answers to the open questions.

Nevertheless I've picked the changeset into my monthly build (currently Kernel 6.1.60, MCU gd32-rev23). Firmware interaction and upgrade works as expected without any additional changes required.

elkablo commented:
I've marked this as WIP because I noticed one potential problem.

M95D commented:
I've been using this for 2 months and it works fine.
What's the status of this PR, and what is the potential problem?

stklcode commented:
This PR backports v4 of the upstream proposal. Current revision is v10 (2024-05-10) with quite a few changes: 
https://patchwork.kernel.org/project/linux-soc/cover/20240510101819.13551-1-kabel@kernel.org/

elkablo commented:
I need to work on upstreaming the driver into kernel, and I am too busy even for that now. I will return to this PR once I have more time.

elkablo committed:
mvebu: 6.6: Backport Turris Omnia MCU patches from 6.11

This backports patches
  dt-bindings: firmware: add cznic,turris-omnia-mcu binding
  platform: cznic: Add preliminary support for Turris Omnia MCU
  platform: cznic: turris-omnia-mcu: Add support for MCU connected GPIOs
  platform: cznic: turris-omnia-mcu: Add support for poweroff and wakeup
  platform: cznic: turris-omnia-mcu: Add support for MCU watchdog
  platform: cznic: turris-omnia-mcu: Add support for MCU provided TRNG
  ARM: dts: turris-omnia: Add MCU system-controller node
  ARM: dts: turris-omnia: Add GPIO key node for front button
  platform: cznic: turris-omnia-mcu: Depend on OF
  platform: cznic: turris-omnia-mcu: Depend on WATCHDOG
  platform: cznic: turris-omnia-mcu: fix Kconfig dependencies
that will be released in 6.11 into mvebu/patches-6.6.

Signed-off-by: Marek Behún <kabel@kernel.org>
Link: https://github.com/openwrt/openwrt/pull/13799
Signed-off-by: Robert Marko <robimarko@gmail.com>

elkablo commented:
The driver has been merged upstream and will be released in Linux 6.11.

I've backported it to 6.6 now and also updated the MCU firmware version and `omnia-mcutool` utility version.

Once CI tests pass, I will unmark the WIP tag.

elkablo committed:
gpio-button-hotplug: add vendor button handling

Handle the KEY_VENDOR key in gpio-button-hotplug driver. This is used
by Turris Omnia.

Signed-off-by: Marek Behún <kabel@kernel.org>
Link: https://github.com/openwrt/openwrt/pull/13799
Signed-off-by: Robert Marko <robimarko@gmail.com>

stklcode commented:
Picked the current status into my monthly build.

Firmware update with`omnia-mcutool --upgrade` stuck in bootloader with "I2C write failed" flashing all LEDs green.
```
# omnia-mcutool --upgrade
Unbound LEDs driver
Unbound front button input driver
Unbound MCU driver
Switching MCU to bootloader...
File /usr/share/omnia-mcu-firmware/gd32-rev23.app.bin (17612 B):
Using old flashing protocol to flash MCU application.
Sending data...
w
omnia-mcutool: I2C write operation failed
```
After a reboot (still flashing, but white) I was able to complete the update 

```
omnia-mcutool -f /tmp/gd32-rev23.app.bin
omnia-mcutool: WARNING: MCU is executing old version of bootloader, cannot determine image validity!
File /tmp/gd32-rev23.app.bin (17612 B):
Using old flashing protocol to flash MCU application.
Sending data...
wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww
wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww
wwwwwwwwww
Receiving data for comparison...
rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr
rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr
rrrrrrrrrr
Read back buffer comparison successful.
Confirming success to MCU... w
MCU's application firmware flashed successfuly. Please reboot!
```

and now it seems to work fine.

```
[   12.580644] turris-omnia-mcu 1-002a: MCU type GD32
[   12.589294] turris-omnia-mcu 1-002a: MCU bootloader firmware version hash: 2174b570bd5c038e83e75026a30eb7250bce7ccc
[   12.601998] turris-omnia-mcu 1-002a: MCU application firmware version hash: d57b774608515e37c2384477d24a4d60e873ecf1
[   12.615976] turris-omnia-mcu 1-002a: registered as rtc1
```

Documented controls behave as expected. LGTM

elkablo commented:
@stklcode so it failed the first time, and you had to manually continue the upgrade. I have flashed the same firmware binary into the same board revision to try to reproduce the issue, but have not been able to. Maybe it was one time thing, but I think that I should make an aborted upgrade somehow recoverable...

Also when trying to reproduce the issue I discovered that the new application firmware cannot be upgraded again when the bootloader firmware is left at the original version. If you have already upgraded to this firmware (which you have), you will need to also upgrade bootloader firmware in order to be able to upgrade application firmware again. I will fix this in new versions so that it does not happen.

For now, you can leave it be, you have newest application firmware. But if you want, you can upgrade bootloader firmware with
```
omnia-mcutool -B -f /usr/share/omnia-mcu-firmware/gd32-rev23.boot.bin
```
This should be very fast, under 1 second. But if you manage to abort it, you will brick your board and will either need to buy JTAG/SWD debugger or send it to Turris Support.

stklcode commented:
> I discovered that the new application firmware cannot be upgraded again when the bootloader firmware is left at the original version.

I think that might have been the issue here, too. Looking through the history I apparently have flashed 3.4 when testing this package a few months ago, so it was the second upgrade with the original old bootloader.

> But if you want, you can upgrade bootloader firmware

Done,  without issues :+1:
I do have a suitable debugger on hand, although never had to use it on the Turris before.

Now boot and app versions are the same and every command I tried seems to work fine so far.

robimarko commented:
Thanks! Rebased on top of main and merged!

drujd commented:
@elkablo So testing this I have found out that the mcu kernel mod is not compiled due to missing CZNIC_PLATFORMS kconf. That HAS to be an oversight, right? Every other kconf regarding that is enabled by default - but unused, due to missing setting of CZNIC_PLATFORMS...

robimarko commented:
@drujd I have tested this before commiting and the kmod is setting `CONFIG_CZNIC_PLATFORMS=y` and its being built

drujd commented:
Ah, OK, it works on clean build. When just rebuilding the last few commits and having kernel built from before, it didn't work (the kmod was not even present in the kernel modules/other menu). Probably will fix itself on next kernel update...

elkablo commented:
@drujd you probably just need to remove the `tmp` directory and run `make menuconfig` again so that the Kconfig options are regenerated.

elkablo commented:
@stklcode PR #16086 bumps `omnia-mcutool` to 0.3-rc3. This should fix the unclean upgrade issue that occurred on your board, and also allows for fixing an aborted upgrade by calling `--upgrade` again.

drujd commented:
Seems to work fine on my crowdfunded Omnia (used it to upgrade stm32-rev23-user-regulator app+boot).
